compiler/rustc_metadata/src/rmeta/encoder.rs

```diff
@@ -716,6 +716,82 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
 }
 
 impl EncodeContext<'a, 'tcx> {
+    fn encode_def_ids(&mut self) {
+        if self.is_proc_macro {
+            return;
+        }
+        let tcx = self.tcx;
+        let hir = tcx.hir();
+        for local_id in hir.iter_local_def_id() {
+            let def_id = local_id.to_def_id();
+            let def_kind = tcx.opt_def_kind(local_id);
+            let def_kind = if let Some(def_kind) = def_kind { def_kind } else { continue };
+            record!(self.tables.def_kind[def_id] <- match def_kind {
+                // Replace Ctor by the enclosing object to avoid leaking details in children crates.
+                DefKind::Ctor(CtorOf::Struct, _) => DefKind::Struct,
+                DefKind::Ctor(CtorOf::Variant, _) => DefKind::Variant,
+                def_kind => def_kind,
+            });
+            record!(self.tables.span[def_id] <- tcx.def_span(def_id));
+            record!(self.tables.attributes[def_id] <- tcx.get_attrs(def_id));
+            record!(self.tables.expn_that_defined[def_id] <- self.tcx.expansion_that_defined(def_id));
+            if let DefKind::Mod
+            | DefKind::Struct
+            | DefKind::Union
+            | DefKind::Enum
+            | DefKind::Variant
+            | DefKind::Trait
+            | DefKind::TyAlias
+            | DefKind::ForeignTy
+            | DefKind::TraitAlias
+            | DefKind::AssocTy
+            | DefKind::Fn
+            | DefKind::Const
+            | DefKind::Static
+            | DefKind::Ctor(..)
+            | DefKind::AssocFn
+            | DefKind::AssocConst
+            | DefKind::Macro(..)
+            | DefKind::Use
+            | DefKind::ForeignMod
+            | DefKind::OpaqueTy
+            | DefKind::Impl
+            | DefKind::Field = def_kind
+            {
+                record!(self.tables.visibility[def_id] <- self.tcx.visibility(def_id));
+            }
+            if let DefKind::Mod
+            | DefKind::Ctor(..)
+            | DefKind::Variant
+            | DefKind::Field
+            | DefKind::Struct
+            | DefKind::AssocTy
+            | DefKind::AssocFn
+            | DefKind::AssocConst
+            | DefKind::TyParam
+            | DefKind::ConstParam
+            | DefKind::Static
+            | DefKind::Const
+            | DefKind::Fn
+            | DefKind::ForeignMod
+            | DefKind::GlobalAsm
+            | DefKind::TyAlias
+            | DefKind::OpaqueTy
+            | DefKind::Enum
+            | DefKind::Union
+            | DefKind::Impl
+            | DefKind::Trait
+            | DefKind::TraitAlias
+            | DefKind::Macro(..)
+            | DefKind::ForeignTy = def_kind
+            {
+                self.encode_stability(def_id);
+                self.encode_const_stability(def_id);
+                self.encode_deprecation(def_id);
```

Comment:
Similarly here, filtering by `DefKind` is a bit fragile as you noted, but then again, not much previous from the previous visitor scheme, where we could forget things everywhere. Maybe just put the big match into a separate function so this code path is less noisy?